### ISSUE TITLE
Allow OperationContract with multiple arguments

### ISSUE THREAD

**Issue by lanfeust69** [protobuf-net/protobuf-net.Grpc#38]
I'm trying to evaluate a replacement of WCF with gRPC, within a big code base with interfaces as "code first" approach. So your work here is very highly appreciated !

So far, I've had decent results, but aside from #27, the biggest challenge for using the exact same interfaces is that we often have operations that have multiple arguments. I can see a number of workarounds, but I was wondering whether it could make sense to handle this case inside protobuf-net.Grpc, basically stuffing all arguments (except a last `CallContext`) in a Tuple (or Tuple-like). I'll try to give it a shot, but would be interested in your thoughts about it.

**Comment by mgravell** [comment]
Yup; agree there is scope to improve this, but my main internal question is : what happens when you change the number of parameters either *from* one, or *back* to one, as changing from:

```
message Foo {...}
```

to

```
message InvisibleWrapper {
    Foo foo = 1;
    Bar bar = 2;
}
message Foo {}

**Comment by lanfeust69** [comment]
Here is what I'm trying (and so far it works well on my use-cases) : in [ContractOperation](https://github.com/protobuf-net/protobuf-net.Grpc/blob/master/src/protobuf-net.Grpc/Internal/ContractOperation.cs#L173), if there is more that one argument of `TypeCategory.Data`, I jam all of these into a single one, a `Tuple<>` with appropriate elements. I hope I'll be able to show something more concrete, but although our company officially encourages contributing to OSS, there are still a few hoops to jump through...

Regarding the breaking changes, I'm not sure I get your point : isn't it the same kind of problem if you change the type of the single Request object of an operation ?

**Comment by mgravell** [comment]
> isn't it the same kind of problem if you change the type of the single Request object of an operation ?

perhaps; and perhaps it is a non-issue, that comes under the heading of "yeah, don't add/remove parameters and expect nothing to break". So; perhaps literally doing nothing except wrapping it is fine.

**Comment by mtsfoni** [comment]
I am in the same place @lanfeust69 was 4 years ago (Yeah, we still run WCF and .net framework in 2024...).

I wonder, did you find a solution for this and #27 that you can share? I want to be able to make an interface in whatever way we deem it right, and use this interface in the client. So "Business-Logic developers" do not need to consider the quirks of the network technology or the related package (When building a prototype for our future stack, I just stumbled over getting an notSupportedException when I have two parameters). Also, I believe true code-first should not force the developer into creating the interface in a specific way.

So if there is no other solution for this, and my teams decides to use C# in our frontend, I might consider making a library that wraps this library. It could work like this:

- The developer would create an interface and gives it an Attribute [GRPCWhatever].
- A similar function to `MapGrpcService<T>()` is being made, that somehow creates the [ServiceContract]-Interface with the known limitation and maps the [GRPCWhatever] calls to the [ServiceContract] interface.
- In the client, something similar could be done in `CreateGrpcService<TService>`
- So the protobuf-net library still would only communicate with the usual ServiceContracts and their limitations, while the developer can freely use multiple parameter, no return values and base types.

@mgravell I would like to get your estimate if that approach is feasible in your eyes, or if I am missing something huge.

I think an extension like that, could make this library extremely attractive for project that communicate between c# programmes.
